File: vfmd_system/src/field_constructor.py
"""
VFMD Simplified Field Constructor
Starts with price velocity field only for Phase 1
"""

import numpy as np
from scipy.ndimage import gaussian_filter
import pandas as pd


class SimpleFieldConstructor:
    """
    Builds Market Vector Field from price data
    Phase 1: Price velocity field only (simplified)
    """
    
    def __init__(self, 
                 spatial_bins=50,
                 temporal_window=100,
                 smoothing_sigma=2.0):
        """
        Args:
            spatial_bins: Number of price levels (grid resolution)
            temporal_window: How many bars to include in field
            smoothing_sigma: Gaussian smoothing parameter
        """
        self.spatial_bins = spatial_bins
        self.temporal_window = temporal_window
        self.smoothing_sigma = smoothing_sigma
        
    def construct_field(self, prices):
        """
        Build vector field from price series
        
        Args:
            prices: Array of prices (most recent = last element)
            
        Returns:
            field: Array of shape (spatial_bins, temporal_bins, 2)
                   where [:, :, 0] = x-component (price direction)
                         [:, :, 1] = y-component (temporal momentum)
        """
        if len(prices) < self.temporal_window:
            raise ValueError(f"Need at least {self.temporal_window} prices")
        
        # Use most recent window
        recent_prices = prices[-self.temporal_window:]
        
        # Normalize prices to [0, 1] for spatial binning
        price_min = np.min(recent_prices)
        price_max = np.max(recent_prices)
        price_range = price_max - price_min
        
        if price_range == 0:
            price_range = 1.0  # Avoid division by zero
        
        normalized_prices = (recent_prices - price_min) / price_range
        
        # Calculate price velocity (change per bar)
        price_velocity = np.diff(recent_prices)
        price_velocity = np.append(price_velocity, price_velocity[-1])  # Pad last
        
        # Calculate price acceleration (change in velocity)
        price_accel = np.diff(price_velocity)
        price_accel = np.append(price_accel, price_accel[-1])
        
        # Create 2D grid: (price_level, time)
        field = np.zeros((self.spatial_bins, self.temporal_window, 2))
        
        # For each time point, map price to spatial bin
        for t in range(self.temporal_window):
            # Which spatial bin does this price belong to?
            price_bin = int(normalized_prices[t] * (self.spatial_bins - 1))
            price_bin = np.clip(price_bin, 0, self.spatial_bins - 1)
            
            # Assign velocity as vector
            # X-component: price velocity (direction of price movement)
            # Y-component: temporal momentum (are we accelerating?)
            field[price_bin, t, 0] = price_velocity[t]
            field[price_bin, t, 1] = price_accel[t]
        
        # Smooth the field to create continuous flow
        for component in range(2):
            field[:, :, component] = gaussian_filter(
                field[:, :, component],
                sigma=self.smoothing_sigma
            )
        
        # Store metadata for later use
        self.price_min = price_min
        self.price_max = price_max
        self.recent_prices = recent_prices
        
        return field
    
    def get_current_position(self):
        """Get grid position of most recent price"""
        if not hasattr(self, 'recent_prices'):
            return None
            
        current_price = self.recent_prices[-1]
        normalized = (current_price - self.price_min) / (self.price_max - self.price_min)
        current_bin = int(normalized * (self.spatial_bins - 1))
        
        return current_bin, self.temporal_window - 1


class FieldAnalyzer:
    """Extract physics quantities from the field"""
    
    @staticmethod
    def compute_gradient_magnitude(field):
        """
        Compute |∇F| across the field
        This shows regions of high force concentration
        """
        # Gradient in each direction
        grad_x = np.gradient(field[:, :, 0], axis=0)
        grad_y = np.gradient(field[:, :, 0], axis=1)
        grad_vx = np.gradient(field[:, :, 1], axis=0)
        grad_vy = np.gradient(field[:, :, 1], axis=1)
        
        # Magnitude of gradient
        magnitude = np.sqrt(grad_x**2 + grad_y**2 + grad_vx**2 + grad_vy**2)
        
        return magnitude
    
    @staticmethod
    def compute_divergence(field):
        """
        Compute ∇·F (divergence)
        Positive = source (accumulation)
        Negative = sink (distribution)
        """
        div_x = np.gradient(field[:, :, 0], axis=0)
        div_y = np.gradient(field[:, :, 1], axis=1)
        
        divergence = div_x + div_y
        return divergence
    
    @staticmethod
    def compute_curl(field):
        """
        Compute curl (vorticity in 2D)
        High values = rotational/choppy flow
        """
        # ∂Fy/∂x - ∂Fx/∂y
        dFy_dx = np.gradient(field[:, :, 1], axis=0)
        dFx_dy = np.gradient(field[:, :, 0], axis=1)
        
        curl = dFy_dx - dFx_dy
        return curl
    
    @staticmethod
    def get_local_vector(field, spatial_pos, temporal_pos):
        """Get vector at specific grid position"""
        if spatial_pos < 0 or spatial_pos >= field.shape[0]:
            return np.array([0, 0])
        if temporal_pos < 0 or temporal_pos >= field.shape[1]:
            return np.array([0, 0])
            
        return field[spatial_pos, temporal_pos, :]


# Test and validation functions
def test_field_on_trend():
    """Test that field detects trending movement"""
    # Create synthetic trending price
    trend = np.linspace(100, 150, 100)  # Uptrend
    noise = np.random.normal(0, 0.5, 100)
    prices = trend + noise
    
    # Build field
    constructor = SimpleFieldConstructor(spatial_bins=50, temporal_window=100)
    field = constructor.construct_field(prices)
    
    # Analyze
    analyzer = FieldAnalyzer()
    divergence = analyzer.compute_divergence(field)
    
    # In uptrend, recent divergence should be positive (accumulation)
    recent_div = divergence[:, -10:].mean()
    
    print(f"Trend test - Recent divergence: {recent_div:.4f}")
    print(f"✅ PASS" if recent_div > 0 else "❌ FAIL: Expected positive divergence")
    
    return field, divergence


def test_field_on_chop():
    """Test that field detects choppy movement"""
    # Create synthetic choppy price
    base = 100
    chop = base + 2 * np.sin(np.linspace(0, 4*np.pi, 100))
    noise = np.random.normal(0, 0.3, 100)
    prices = chop + noise
    
    # Build field
    constructor = SimpleFieldConstructor(spatial_bins=50, temporal_window=100)
    field = constructor.construct_field(prices)
    
    # Analyze
    analyzer = FieldAnalyzer()
    curl = analyzer.compute_curl(field)
    
    # In chop, curl should be elevated (rotational flow)
    avg_curl = np.abs(curl).mean()
    
    print(f"Chop test - Average curl: {avg_curl:.4f}")
    print(f"✅ PASS" if avg_curl > 0.01 else "❌ FAIL: Expected higher curl in chop")
    
    return field, curl


if __name__ == "__main__":
    print("Testing Field Constructor...")
    print("=" * 50)
    
    print("\n1. Testing on trending data:")
    trend_field, trend_div = test_field_on_trend()
    
    print("\n2. Testing on choppy data:")
    chop_field, chop_curl = test_field_on_chop()
    
    print("\n" + "=" * 50)
    print("Basic tests complete. Ready for real data.")

File: server/services/vfmd/fieldConstructor.ts
/**
 * VFMD Field Constructor - TypeScript Port
 * Builds market vector field from price/volume data
 *
 * The field maps price levels (spatial dimension) across time (temporal)
 * with vector components representing velocity and acceleration
 */

import type { MarketTick, VectorField } from './types';

/**
 * Simple Gaussian blur for field smoothing
 */
function gaussianBlur(
  matrix: number[][],
  sigma: number,
  iterations: number = 1
): number[][] {
  if (sigma <= 0) return matrix;

  const height = matrix.length;
  const width = matrix[0].length;
  let result = matrix.map(row => [...row]);

  for (let iter = 0; iter < iterations; iter++) {
    const kernel = createGaussianKernel(sigma);
    const newResult: number[][] = result.map(row => [...row]);

    for (let y = 1; y < height - 1; y++) {
      for (let x = 1; x < width - 1; x++) {
        let sum = 0;
        let weight = 0;

        for (let ky = -1; ky <= 1; ky++) {
          for (let kx = -1; kx <= 1; kx++) {
            const k = kernel[ky + 1][kx + 1];
            sum += result[y + ky][x + kx] * k;
            weight += k;
          }
        }

        newResult[y][x] = sum / weight;
      }
    }

    result = newResult;
  }

  return result;
}

function createGaussianKernel(sigma: number): number[][] {
  // Guard against tiny sigma producing 1×1 kernel → division by zero
  if (sigma < 0.6) {
    return [[1]]; // return identity kernel for very small sigma
  }

  const size = Math.ceil(sigma * 3) * 2 + 1;
  const kernel: number[][] = [];
  const mean = Math.floor(size / 2);
  let sum = 0;

  for (let i = 0; i < size; i++) {
    kernel[i] = [];
    for (let j = 0; j < size; j++) {
      const x = i - mean;
      const y = j - mean;
      const val =
        Math.exp(-(x * x + y * y) / (2 * sigma * sigma)) /
        (2 * Math.PI * sigma * sigma);
      kernel[i][j] = val;
      sum += val;
    }
  }

  // Normalize (sum > 0 guaranteed due to guard above)
  if (sum > 0) {
    for (let i = 0; i < size; i++) {
      for (let j = 0; j < size; j++) {
        kernel[i][j] /= sum;
      }
    }
  }

  return kernel;
}

/**
 * Construct vector field from price data
 */
export class FieldConstructor {
  private spatialBins: number;
  private temporalWindow: number;
  private smoothingSigma: number;

  constructor(
    spatialBins: number = 50,
    temporalWindow: number = 100,
    smoothingSigma: number = 2.0
  ) {
    this.spatialBins = spatialBins;
    this.temporalWindow = temporalWindow;
    this.smoothingSigma = smoothingSigma;
  }

  /**
   * Build vector field from price series
   * Returns field of shape [spatialBins, temporalWindow, 2]
   */
  constructField(prices: number[]): VectorField {
    if (prices.length < this.temporalWindow) {
      throw new Error(
        `Need at least ${this.temporalWindow} prices, got ${prices.length}`
      );
    }

    // Use most recent window
    const recentPrices = prices.slice(-this.temporalWindow);

    // Normalize prices to [0, 1]
    const priceMin = Math.min(...recentPrices);
    const priceMax = Math.max(...recentPrices);
    let priceRange = priceMax - priceMin;

    if (priceRange === 0) {
      priceRange = 1.0;
    }

    const normalizedPrices = recentPrices.map(p => (p - priceMin) / priceRange);

    // Calculate price velocity (change per bar) — NORMALIZED BY PRICE RANGE
    // Velocities are now in returns-space (% of window range) instead of raw dollar-space
    // This makes them scale-agnostic: BTC and ETH velocities are now comparable
    const priceVelocity: number[] = [];
    for (let i = 1; i < recentPrices.length; i++) {
      priceVelocity.push((recentPrices[i] - recentPrices[i - 1]) / priceRange);
    }
    priceVelocity.push(priceVelocity[priceVelocity.length - 1] || 0);

    // Calculate acceleration (already normalized since velocities are normalized)
    const priceAccel: number[] = [];
    for (let i = 1; i < priceVelocity.length; i++) {
      priceAccel.push(priceVelocity[i] - priceVelocity[i - 1]);
    }
    priceAccel.push(priceAccel[priceAccel.length - 1] || 0);

    // Create 3D field [spatialBins, temporalWindow, 2]
    const field: number[][][] = Array(this.spatialBins)
      .fill(null)
      .map(() =>
        Array(this.temporalWindow)
          .fill(null)
          .map(() => [0, 0])
      );

    // Populate field: for each time point, map price to spatial bin
    for (let t = 0; t < this.temporalWindow; t++) {
      const priceBin = Math.min(
        this.spatialBins - 1,
        Math.max(0, Math.floor(normalizedPrices[t] * (this.spatialBins - 1)))
      );

      // X: price velocity, Y: acceleration
      field[priceBin][t][0] = priceVelocity[t];
      field[priceBin][t][1] = priceAccel[t];
    }

    // Smooth each component separately
    for (let component = 0; component < 2; component++) {
      const slice: number[][] = field.map(spatial => spatial.map(t => t[component]));
      const smoothed = gaussianBlur(slice, this.smoothingSigma, 1);

      for (let s = 0; s < this.spatialBins; s++) {
        for (let t = 0; t < this.temporalWindow; t++) {
          field[s][t][component] = smoothed[s][t];
        }
      }
    }

    return {
      data: field,
      spatialBins: this.spatialBins,
      temporalWindow: this.temporalWindow,
      priceMin,
      priceMax
    };
  }

  /**
   * Get current position of most recent price in the grid
   */
  getCurrentPosition(field: VectorField): [number, number] {
    const lastPrice = field.priceMax; // Approximate
    const normalized =
      (lastPrice - field.priceMin) /
      (field.priceMax - field.priceMin || 1);
    const spatialBin = Math.min(
      field.spatialBins - 1,
      Math.max(0, Math.floor(normalized * (field.spatialBins - 1)))
    );

    return [spatialBin, field.temporalWindow - 1];
  }
}

/**
 * Extract physics quantities from field
 */
export class FieldAnalyzer {
  /**
   * Compute gradient magnitude |∇F|
   * Shows regions of high force concentration
   */
  static computeGradientMagnitude(field: VectorField): number[][] {
    const data = field.data;
    const s = field.spatialBins;
    const t = field.temporalWindow;

    const magnitude: number[][] = Array(s)
      .fill(null)
      .map(() => Array(t).fill(0));

    for (let i = 0; i < s; i++) {
      for (let j = 0; j < t; j++) {
        // Gradient in spatial direction (i)
        const grad_i_x =
          i < s - 1 ? data[i + 1][j][0] - data[i][j][0] : 0;
        const grad_i_y =
          i < s - 1 ? data[i + 1][j][1] - data[i][j][1] : 0;

        // Gradient in temporal direction (j)
        const grad_j_x =
          j < t - 1 ? data[i][j + 1][0] - data[i][j][0] : 0;
        const grad_j_y =
          j < t - 1 ? data[i][j + 1][1] - data[i][j][1] : 0;

        magnitude[i][j] = Math.sqrt(
          grad_i_x * grad_i_x +
            grad_i_y * grad_i_y +
            grad_j_x * grad_j_x +
            grad_j_y * grad_j_y
        );
      }
    }

    return magnitude;
  }

  /**
   * Compute divergence ∇·F
   * Positive = source (accumulation), Negative = sink (distribution)
   */
  static computeDivergence(field: VectorField): number[][] {
    const data = field.data;
    const s = field.spatialBins;
    const t = field.temporalWindow;

    const divergence: number[][] = Array(s)
      .fill(null)
      .map(() => Array(t).fill(0));

    for (let i = 0; i < s; i++) {
      for (let j = 0; j < t; j++) {
        // ∂Fx/∂i (spatial gradient of x-component)
        const div_x =
          i < s - 1 ? data[i + 1][j][0] - data[i][j][0] : 0;

        // ∂Fy/∂j (temporal gradient of y-component)
        const div_y =
          j < t - 1 ? data[i][j + 1][1] - data[i][j][1] : 0;

        divergence[i][j] = div_x + div_y;
      }
    }

    return divergence;
  }

  /**
   * Compute curl (vorticity in 2D)
   * High values = rotational/choppy flow
   */
  static computeCurl(field: VectorField): number[][] {
    const data = field.data;
    const s = field.spatialBins;
    const t = field.temporalWindow;

    const curl: number[][] = Array(s)
      .fill(null)
      .map(() => Array(t).fill(0));

    for (let i = 0; i < s; i++) {
      for (let j = 0; j < t; j++) {
        // ∂Fy/∂i
        const dFy_di =
          i < s - 1 ? data[i + 1][j][1] - data[i][j][1] : 0;

        // ∂Fx/∂j
        const dFx_dj =
          j < t - 1 ? data[i][j + 1][0] - data[i][j][0] : 0;

        curl[i][j] = dFy_di - dFx_dj;
      }
    }

    return curl;
  }

  /**
   * Get vector at specific grid position
   */
  static getLocalVector(
    field: VectorField,
    spatialPos: number,
    temporalPos: number
  ): [number, number] {
    if (
      spatialPos < 0 ||
      spatialPos >= field.spatialBins ||
      temporalPos < 0 ||
      temporalPos >= field.temporalWindow
    ) {
      return [0, 0];
    }

    return field.data[spatialPos][temporalPos] as [number, number];
  }
}
